server/src/main/java/org/elasticsearch/search/aggregations/bucket/terms/SignificantTextAggregatorFactory.java

```diff
@@ -162,12 +157,120 @@ protected Aggregator createInternal(Aggregator parent, CardinalityUpperBound car
         );
     }
 
+    /**
+     * Create the {@link CollectorSource}, gathering some timing information
+     * if we're profiling.
+     * <p>
+     * When profiling aggregations {@link LeafBucketCollector#collect(int, long)} method
+     * out of the box but our implementation of that method does three things that is
+     * useful to get timing for:
+     * <ul>
+     * <li>Fetch field values from {@code _source}
+     * <li>Analyze the field
+     * <li>Do all the normal {@code terms} agg stuff with its terms
+     * </ul>
+     * <p>
+     * The most convenient way to measure all of these is to time the fetch and all
+     * the normal {@code terms} agg stuff. You can then subtract those timings from
+     * the overall collect time to get the analyze time. You can also get the total
+     * number of terms that we analyzed by looking at the invocation count on the
+     * {@code terms} agg stuff.
+     * <p>
+     * While we're at it we count the number of values we fetch from source.
+     */
+    private CollectorSource createCollectorSource() {
+        Analyzer analyzer = context.getIndexAnalyzer(f -> {
+            throw new IllegalArgumentException("No analyzer configured for field " + f);
+        });
+        if (context.profiling()) {
+            return new SignificantTextCollectorSource(
+                context.lookup().source(),
+                context.bigArrays(),
+                fieldType,
+                analyzer,
+                sourceFieldNames,
+                filterDuplicateText
+            ) {
+                private final Timer extract = new Timer();
+                private final Timer collectAnalyzed = new Timer();
+                private long valuesFetched;
+                private long charsFetched;
+
+                @Override
+                protected void processTokenStream(
+                    StringFilter includeExclude,
+                    int doc,
+                    long owningBucketOrd,
+                    String text,
+                    TokenStream ts,
+                    BytesRefHash inDocTerms,
+                    DuplicateByteSequenceSpotter spotter,
+                    LongConsumer addRequestCircuitBreakerBytes,
+                    LeafBucketCollector sub,
+                    CollectConsumer consumer
+                ) throws IOException {
+                    valuesFetched++;
+                    charsFetched += text.length();
```

Comment:
With these two counters and the `Timer` around extract we can tell if the we are extracting many fields or a few fields. We can see if we're extracting many fields and filtering them away. And we can see how long the fields are in utf-16 `char`s. Which is better than nothing. Not as good as bytes, but such is life.